### PR TITLE
Basic geometric twap e2e test

### PR DESCRIPTION
**Pull request by hieuvubk**
Closes: #3728 

## What is the purpose of the change

This PR adds a geometric twap e2e test.
It does the following:  creates a pool, queries twap, performs a swap, and queries twap again.

Twap is expected to change after the swap.
The pool is created with 1_000_000 uosmo and 2_000_000 stake and equal weights.

Assuming base asset is uosmo, the initial twap is 2

Upon swapping 1_000_000 uosmo for uatom, supply changes, making uosmo less expensive.

As a result of the swap, twap changes to 0.5.

## Testing and Verifying

This change added tests.

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not applicable

### CODE REVIEW

File: tests/e2e/e2e_test.go [hunk 1]
```diff
@@ -569,3 +569,94 @@ func (s *IntegrationTestSuite) TestExpeditedProposals() {
 	s.T().Logf("expeditedVotingPeriodDuration within two seconds of expected time: %v", timeDelta)
 	close(totalTimeChan)
 }
+
+// TestGeometricTWAP tests geometric twap.
+// It does the following:  creates a pool, queries twap, performs a swap ,and queries twap again.
```
Comment (p0mvn) [inline tests/e2e/e2e_test.go:574]:
```suggestion
// It does the following:  creates a pool, queries twap, performs a swap , and queries twap again.
```

File: tests/e2e/e2e_test.go [hunk 2]
```diff
@@ -569,3 +569,94 @@ func (s *IntegrationTestSuite) TestExpeditedProposals() {
...
+	s.Require().NoError(err)
+
+	// We swap uosmo so uosmo's supply will increase and stake will decrease.
+	// The the price after will be smaller than the previous one.
+	s.Require().True(initialTwapBOverA.GT(afterSwapTwapBOverA))
+
+	// Assume base = uosmo, quote = stake
+	// At pool creation, we had:
+	// quote assset supply / base asset supply = 2_000_000 / 1_000_000 = 2
+	// Next, we swapped 1_000_000 usomo for uatom.
```
Comment (p0mvn) [inline tests/e2e/e2e_test.go:656]:
```suggestion
	// Next, we swapped 1_000_000 uosmo for uatom.
```

**Review by p0mvn — APPROVED** [review]
LGTM. Thank you for implementing this e2e test @hieuvubk 

I added some comments to explain what's going on here and removed some logic that is not needed for testing geometric twap specifically. For example, made swap and exit fee be zero and weights be 1 for simpler calculations.

The core logic of the test is sound, great work!

File: tests/e2e/configurer/chain/queries.go [hunk 3]
```diff
@@ -209,13 +208,53 @@ func (n *NodeConfig) QueryArithmeticTwap(poolId uint64, baseAsset, quoteAsset st
...
+		"quote_asset", quoteAsset,
+		"start_time", startTime.Format(time.RFC3339Nano),
+	)
+	if err != nil {
+		return sdk.Dec{}, err
+	}
+
+	var response twapqueryproto.GeometricTwapToNowResponse
+	err = util.Cdc.UnmarshalJSON(bz, &response)
+	require.NoError(n.t, err) // this error should not happen
```
Comment (czarcas7ic) [inline tests/e2e/configurer/chain/queries.go:233]:
```suggestion
	require.NoError(n.t, err)
```
nit

File: tests/e2e/configurer/chain/queries.go [hunk 4]
```diff
@@ -209,13 +208,53 @@ func (n *NodeConfig) QueryArithmeticTwap(poolId uint64, baseAsset, quoteAsset st
...
+		"start_time", startTime.Format(time.RFC3339Nano),
+		"end_time", endTime.Format(time.RFC3339Nano),
+	)
+	if err != nil {
+		return sdk.Dec{}, err
+	}
+
+	var response twapqueryproto.GeometricTwapResponse
+	err = util.Cdc.UnmarshalJSON(bz, &response)
+	require.NoError(n.t, err) // this error should not happen
```
Comment (czarcas7ic) [inline tests/e2e/configurer/chain/queries.go:254]:
```suggestion
	require.NoError(n.t, err)
```

File: tests/e2e/e2e_test.go [hunk 5]
```diff
@@ -569,3 +569,94 @@ func (s *IntegrationTestSuite) TestExpeditedProposals() {
 	s.T().Logf("expeditedVotingPeriodDuration within two seconds of expected time: %v", timeDelta)
 	close(totalTimeChan)
 }
+
+// TestGeometricTWAP tests geometric twap.
+// It does the following:  creates a pool, queries twap, performs a swap , and queries twap again.
+// Twap is expected to change after the swap.
+// The pool is created with 1_000_000 uosmo and 2_000_000 stake and equal weights.
+// Assuming base asset is uosmo, the initial twap is 2
+// Upon swapping 1_000_000 uosmo for uatom, supply changes, making uosmo less expensive.
```
Comment (czarcas7ic) [inline tests/e2e/e2e_test.go:578]:
```suggestion
// Upon swapping 1_000_000 uosmo for stake, supply changes, making uosmo less expensive.
```

File: tests/e2e/e2e_test.go [hunk 6]
```diff
@@ -569,3 +569,94 @@ func (s *IntegrationTestSuite) TestExpeditedProposals() {
...
+	initialTwapAOverB, err := chainANode.QueryGeometricTwapToNow(poolId, denomB, denomA, timeBeforeSwapPlus5ms)
+	s.Require().NoError(err)
+	s.Require().Equal(sdk.NewDecWithPrec(5, 1), initialTwapAOverB)
+
+	coinAIn := fmt.Sprintf("1000000%s", denomA)
+	chainANode.BankSend(coinAIn, chainA.NodeConfigs[0].PublicAddress, swapWalletAddr)
+
+	s.T().Logf("performing swap of %s for %s", coinAIn, denomB)
+
+	// uatom out = uatom supply * (1 - (uosmo supply before / uosmo supply after)^(uosmo weight / uatom weight))
```
Comment (czarcas7ic) [inline tests/e2e/e2e_test.go:631]:
```suggestion
	// stake out = stake supply * (1 - (uosmo supply before / uosmo supply after)^(uosmo weight / stake weight))
```

File: tests/e2e/e2e_test.go [hunk 7]
```diff
@@ -569,3 +569,94 @@ func (s *IntegrationTestSuite) TestExpeditedProposals() {
...
+	//           = 2_000_000 * (1 - (1_000_000 / 2_000_000)^1)
+	//           = 2_000_000 * 0.5
+	//           = 1_000_000
+	chainANode.SwapExactAmountIn(coinAIn, minAmountOut, fmt.Sprintf("%d", poolId), denomB, swapWalletAddr)
+
+	// New supply post swap:
+	// uatom = 2_000_000 - 1_000_000 - 1_000_000
```
Comment (czarcas7ic) [inline tests/e2e/e2e_test.go:638]:
```suggestion
	// stake = 2_000_000 - 1_000_000 - 1_000_000
```

File: tests/e2e/e2e_test.go [hunk 8]
```diff
@@ -569,3 +569,94 @@ func (s *IntegrationTestSuite) TestExpeditedProposals() {
...
+	s.Require().NoError(err)
+
+	// We swap uosmo so uosmo's supply will increase and stake will decrease.
+	// The the price after will be smaller than the previous one.
+	s.Require().True(initialTwapBOverA.GT(afterSwapTwapBOverA))
+
+	// Assume base = uosmo, quote = stake
+	// At pool creation, we had:
+	// quote assset supply / base asset supply = 2_000_000 / 1_000_000 = 2
+	// Next, we swapped 1_000_000 uosmo for uatom.
```
Comment (czarcas7ic) [inline tests/e2e/e2e_test.go:656]:
```suggestion
	// Next, we swapped 1_000_000 uosmo for stake.
```

**Review by czarcas7ic — APPROVED** [review]

**Pull request merged**